src/codegen/solana_accounts/account_collection.rs

```diff
@@ -505,3 +452,79 @@ fn check_expression(expr: &Expression, data: &mut RecurseData) -> bool {
 
     true
 }
+
+fn transfer_accounts(
+    loc: &pt::Loc,
+    contract_no: usize,
+    function_no: usize,
+    data: &mut RecurseData,
+    program_id_present: bool,
+) {
+    let accounts_to_add = data.functions[function_no].solana_accounts.borrow().clone();
+
+    for (name, mut account) in accounts_to_add {
+        if name == BuiltinAccounts::DataAccount {
+            let idl_name = format!("{}_dataAccount", data.contracts[contract_no].name);
+            if let Some(acc) = data.functions[data.ast_no]
+                .solana_accounts
+                .borrow()
+                .get(&idl_name)
+            {
+                if acc.loc != *loc {
+                    data.diagnostics.push(
+                        Diagnostic::error_with_note(
+                            *loc,
+                            format!("contract '{}' is called more than once in this function, so automatic account collection cannot happen. \
+                                         Please, provide the necessary accounts using the {{accounts:..}} call argument", data.contracts[contract_no].name),
```

Comment:
This limitation makes the automatic account collection useless, not sure why we are bothering with this code.

Why can't I call the same function twice? 

The diagnostic says "called" but this includes both constructors and calls.

Do you know how anchor deals with this problem?

Reply:
This is a good point for discussion.

Now that the data account is separated from the contract, it is another parameter we need whenever calling a function that needs it. For the same instance of a contract, I can call functions with different data accounts.
Let `Foo f` be a contract, then:
```
f.bar(); // use one data account
f.bar(); // use another data account
```
I wanted to be cautious to not assume that these calls will use the same data account. This problem exists with any combination of functions from a contract, including a constructor and another function. The constructor only initializes a data account, and calling another function from that contract does not mean we want to use the same account again.

Anchor requires all accounts to be manually passed to a CPI. There is no automatic collection. Anchor, however, abstracts away the signer and writer flags, only requiring developers to specify the address for each of the declared accounts. So, if a function `bar` requires `acc1` and `acc2`, one needs to type `{acc1: address1, acc2: address2}`. Check the example [here](https://www.anchor-lang.com/docs/cross-program-invocations).

The account management is becoming more limited in that there are even more cases in which it does not work. We can get rid of it and follow the same approach as Anchor as soon as we have the account annotations in place. 

Removing the management would simplify our code. If this is the case, I'd rather do this in a separate PR, because it will require breaking changes to what we already have in constructor calls.

Reply:
As per your message on Discord, we'll leave the account management limited.